### PR TITLE
Adding tenancy to the database

### PR DESCRIPTION
**Pull request by nir-sopher**
WIP
On this PR the tenancy is only added to the relavant tables. The already existing APIs were not modified.
Default tenancy of items is "NULL".
API for tenancy was added, including UT.
Additionally, an "upgrade" script was added to create the root tenant and assign it to the specified admin user.

A "add user" API was added, as well as a simple test.

====Second push
Tools were added to check trnancy
CDN/User CRUD APIs were adjusted to include the tenant_id
UI: Tenant of an object is the same as the tenant of the creating user

### CODE REVIEW

File: traffic_ops/app/lib/API/Tenant.pm [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+	my @data;
+	my $orderby = $self->param('orderby') || "name";
+	my $rs_data = $self->db->resultset("Tenant")->search( undef, {order_by => 'me.' . $orderby } );
+	while ( my $row = $rs_data->next ) {
+		push(
+			@data, {
+				"id"           => $row->id,
+				"name"         => $row->name,
+				"active"       => $row->active,
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: I think you have to "cast" the active flag to ensure that true|false shows in the api as opposed to 0|1

Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: so...."active"       => \\$row->active,

File: traffic_ops/app/lib/API/Tenant.pm [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	my $id   = $self->param('id');
+
+	my $rs_data = $self->db->resultset("Tenant")->search( { 'me.id' => $id });
+	my @data = ();
+	while ( my $row = $rs_data->next ) {
+		push(
+			@data, {
+				"id"           => $row->id,
+				"name"         => $row->name,
+				"active"       => $row->active,
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: same thing with active casting

File: traffic_ops/app/t/api/1.2/tenant.t [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+#no_transactions=>1 ==> keep fixtures after every execution, beware of duplicate data!
+#no_transactions=>0 ==> delete fixtures after every execution
+
+BEGIN { $ENV{MOJO_MODE} = "test" }
+
+my $schema = Schema->connect_to_database;
+my $dbh    = Schema->database_handle;
+my $t      = Test::Mojo->new('TrafficOps');
+
+my $false = 0;
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/tenant.t]: I wouldn't use variables. 0 and 1 should be fine.

File: traffic_ops/app/lib/API/Tenant.pm [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+		active    => $params->{active},
+		parent_id => $params->{parentId}
+	};
+
+	my $rs = $tenant->update($values);
+	if ($rs) {
+		my $response;
+		$response->{id}          = $rs->id;
+		$response->{name}        = $rs->name;
+		$response->{active}      = $rs->active;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: ensure this returns true|false in the api

File: traffic_ops/app/lib/API/Tenant.pm [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+		parent_id 	=> $params->{parentId}
+	};
+
+	my $insert = $self->db->resultset('Tenant')->create($values);
+	my $rs = $insert->insert();
+	if ($rs) {
+		my $response;
+		$response->{id}          	= $rs->id;
+		$response->{name}        	= $rs->name;
+		$response->{active}        	= $rs->active;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: ensure this returns true|false in the api

File: traffic_ops/app/lib/API/Tenant.pm [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+		return $self->success( $response, "Tenant create was successful." );
+	}
+	else {
+		return $self->alert("Tenant create failed.");
+	}
+
+}
+
+
+sub delete {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: do you want to check here that you can't delete a tenant that has a child tenant? or check that you can't delete a tenant that is assigned to a ds or a cdn? and return a 400. rather than letting the db return a 500?

File: traffic_ops/app/lib/Schema/Result/GooseDbVersion.pm [hunk 7]
```diff
@@ -25,25 +25,24 @@ __PACKAGE__->table("goose_db_version");
```
Comment (mitchell852) [inline traffic_ops/app/lib/Schema/Result/GooseDbVersion.pm]: not really sure why this table got touched....

File: traffic_ops/app/lib/API/User.pm [hunk 8]
```diff
@@ -218,6 +218,118 @@ sub update {
...
+		new_user 			=> ( $params->{newUser} ) ? 1 : 0,
+		phone_number 			=> defined_or_default($params->{phoneNumber}, ""),
+		postal_code 			=> defined_or_default($params->{postalCode}, ""),
+		public_ssh_key 			=> defined_or_default($params->{publicSshKey}, ""),
+		registration_sent 		=> defined_or_default( $params->{registrationSent}, undef),
+		role 				=> $params->{role},
+		state_or_province 		=> defined_or_default($params->{stateOrProvince}, ""),
+		username 			=> $params->{username},
+		new_user            		=> defined_or_default($params->{newUser}, 0),		
+		uid                  		=> defined_or_default($params->{uid}, 0),
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: I don't think you'd ever want the api consumer to define the uid or gid of the user...so i would just leave it out. if they send it, it gets ignored.

File: traffic_ops/app/lib/API/User.pm [hunk 9]
```diff
@@ -218,6 +218,118 @@ sub update {
...
+		return $self->alert("local-password and confirmed-local-password mismatch.");
+	}
+	
+	if ( !defined($params->{role}) ) {
+		return $self->alert("role is required.");
+	}
+	
+	
+
+	my $values = {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: aren't they supposed to defined the tenantId of the user?

File: traffic_ops/app/t/api/1.2/tenant.t [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+#adding a child inactive tenant
+ok $t->post_ok('/api/1.2/tenants' => {Accept => 'application/json'} => json => {
+        "name" => "tenantE", "active" => $false, "parentId" => $tenantA_id })->status_is(200)->or( sub { diag $t->tx->res->content->asset->{content}; } )
+	->json_is( "/response/name" => "tenantE" )
+	->json_is( "/response/active" => $false )
+	->json_is( "/response/parentId" =>  $tenantA_id)
+            , 'Does the tenant details return?';
+
+#cannot delete a tenant that have children
+ok $t->delete_ok('/api/1.2/tenants/' . $tenantA_id)->status_is(500);
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/tenant.t]: how about returning a 400 instead in the code rather than letting the db throw a 500?

File: traffic_ops/app/lib/API/Tenant.pm (hunk 1 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: ensure this returns a true|false in the api.  you might have to cast it like \\$row->active,

File: traffic_ops/app/lib/API/Tenant.pm (hunk 2 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: ensure this returns a true|false in the api. you might have to cast it like \$row->active,

File: traffic_ops/app/lib/API/Tenant.pm (hunk 4 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: ensure this returns a true|false in the api. you might have to cast it like \$row->active,

File: traffic_ops/app/lib/API/Tenant.pm (hunk 5 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: ensure this returns a true|false in the api. you might have to cast it like \$row->active,

File: traffic_ops/app/lib/API/Tenant.pm (hunk 6 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: can you delete a tenant that:
- has child tenants
- is assigned to a user
- is assigned to a ds
- is assigned to a cdn

if not, you want to throw 400's for that?

File: traffic_ops/app/lib/Schema/Result/GooseDbVersion.pm (hunk 7 above)
Comment (mitchell852) [inline traffic_ops/app/lib/Schema/Result/GooseDbVersion.pm]: not sure why this file was touched...doesn't seem related to your changes

File: traffic_ops/app/t/api/1.2/tenant.t (hunk 3 above)
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/tenant.t]: i would get rid of these variables and just use 0 or 1 instead.

File: traffic_ops/app/lib/API/Tenant.pm (hunk 6 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: No. Protection is done via the DB.
Need to UT that

File: traffic_ops/app/t/api/1.2/tenant.t (hunk 10 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/tenant.t]: Added to my to do list - improve later on

File: traffic_ops/app/lib/API/User.pm (hunk 9 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: This will be added a separate commit

File: traffic_ops/app/lib/API/User.pm (hunk 8 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: OK. Will remove it when I update this PR in the future

File: traffic_ops/app/lib/API/Tenant.pm (hunk 1 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: Will fix

File: traffic_ops/app/lib/API/Tenant.pm (hunk 2 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: \$raw-active again

File: traffic_ops/app/lib/API/Tenant.pm (hunk 4 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: \ again

File: traffic_ops/app/lib/API/Tenant.pm (hunk 5 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: \ again

File: traffic_ops/app/lib/API/Tenant.pm (hunk 6 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: Added to my list - in the future will further test relations instead of counting on the DB

Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: Maybe we should avoid deleting a tenant???

File: traffic_ops/app/t/api/1.2/tenant.t (hunk 3 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/tenant.t]: I prefer it like this for readability

**Comment by nir-sopher** [comment]
TO DO list (10x Jeremy!)
remove uid & gid from the add: will do in a future commit
true\false in the API
Verify no dependancy for a deleted tenant

File: traffic_ops/app/lib/API/Tenant.pm (hunk 6 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: i think deleting a tenant is an important feature. helps with cleanup.

File: traffic_ops/app/t/api/1.2/tenant.t (hunk 3 above)
Comment (dangogh) [inline traffic_ops/app/t/api/1.2/tenant.t]: However,  it's considered bad style in Perl because there are several values for a "false" value (and many more for "true"...).  Just one example:

```
my $false = 0;
if ($self->active == $false) {
}
```
The above will be incorrect if `$self->active` is undef.

File: traffic_ops/app/lib/API/Tenant.pm (hunk 6 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/Tenant.pm]: Ack, will add verifications

File: traffic_ops/app/t/api/1.2/tenant.t (hunk 3 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/tenant.t]: ACK

**Comment by mitchell852** [comment]
Is this still a WIP? Or can it be merged when approved?

**Comment by nir-sopher** [comment]
This PR can stand for its own. It passed the unit tests beside
delivery-service related failures, which as far as I can see are not
related to my change. Will be able to double check only next week.

Next to come are
0. Changes following your remarks
1. Delivery service API - add tenancy
2. Warning on put requests with no tenancy check.
3. Delivery enforcement.

So no much changes in already PR code.

On the other hand, it might be best to pull it after step "2" - just before
we start to enforce stuff.

Nir

File: traffic_ops/app/db/migrations/20170221000001_initial_tenancy.sql [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+ALTER TABLE cdn
+DROP COLUMN tenant_id;
+
+ALTER TABLE tm_user
+DROP COLUMN tenant_id;
+
+DROP TRIGGER on_update_current_timestamp ON tenant;
+
+DROP TABLE tenant;
```
Comment (mitchell852) [inline traffic_ops/app/db/migrations/20170221000001_initial_tenancy.sql]: do you need to drop the 3 indexes you created as well? in the goose down?

File: traffic_ops/app/lib/API/Tenant.pm (hunk 1 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: looks like this casting isnt done yet

File: traffic_ops/app/lib/API/Tenant.pm (hunk 2 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/Tenant.pm]: need to cast boolean

File: traffic_ops/app/lib/API/User.pm [hunk 12]
```diff
@@ -218,6 +226,121 @@ sub update {
...
+		return $self->alert("Username is required.");
+	}
+	
+	my $existing = $self->db->resultset('TmUser')->search( { username => $name } )->single();
+	if ($existing) {
+		return $self->alert("A user with username \"$name\" already exists.");
+	}
+
+
+	if ( !defined($params->{fullName}) ) {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: this is probably not required in my opinion

**Comment by mitchell852** [comment]
@nir-sopher - let me know when you get this done and i'll review again - 

0. Changes following your remarks done
1. Delivery service API - add tenancy

**Comment by nir-sopher** [comment]
These columns are not relevant after downgrade (until next upgrade).
What is the common practice in such cases in TC?

**Comment by nir-sopher** [comment]
The \ is still in my local repo. Will be pushed on next update

On Mar 6, 2017 8:35 PM, "Jeremy Mitchell" <notifications@github.com> wrote:

*@mitchell852* commented on this pull request.
------------------------------

In traffic_ops/app/lib/API/Tenant.pm
<https://github.com/apache/incubator-trafficcontrol/pull/322#discussion_r104487370>
:

> +	my $tenant_id	= shift;
+	return !defined($self->db->resultset('Tenant')->search( { id =>
$tenant_id } )->get_column('parent_id')->single());
+}
+
+sub index {
+	my $self 	= shift;
+
+	my @data;
+	my $orderby = $self->param('orderby') || "name";
+	my $rs_data = $self->db->resultset("Tenant")->search( undef,
{order_by => 'me.' . $orderby } );
+	while ( my $row = $rs_data->next ) {
+		push(
+			@data, {
+				"id"           => $row->id,
+				"name"         => $row->name,
+				"active"       => $row->active,

looks like this casting isnt done yet

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/apache/incubator-trafficcontrol/pull/322#discussion_r104487370>,
or mute the thread

**Comment by nir-sopher** [comment]
Current status:
All UTs passed.
First batch of PR remarks was fixed.
The API supports tenancy for CDn / Delivery Service / Users
UTs are added to cover this functionality.

Next, I will try to break this PR into smaller PRs.

Main gap: tenancy verification upon operation. This is for another PR

**Comment by mitchell852** [comment]
can this PR be closed?

**Comment by nir-sopher** [comment]
Almost all changes are already added by other PRs.
Closing
